Updated `no-dupe-disjunctions`

RunDevelopment opened the pull request:
This brings over the functionality of `clean-regex/disjoint-alternatives`.

---

@ota-meshi This PR is still WIP at the moment. I wanted to ask you a few questions (see my comments) before I continue working on this.

---

The results using Prism's 2k regexes. 

The AST approach (just using `isCoveredNode`) found 25 problems:

<details>

```
C:\Users\micha\Git\prism\components\prism-abap.js
  19:4043  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-actionscript.js
  2:316  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-agda.js
  22:76  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-bash.js
   37:23   warning  Unexpected useless alternative. This alternative is already covered by '--?' and can be removed
   regexp/no-dupe-disjunctions
   37:32   warning  Unexpected useless alternative. This alternative is already covered by '\+\+?' and can be removed
   regexp/no-dupe-disjunctions
   37:48   warning  Unexpected useless alternative. This alternative is already covered by '\*\*?' and can be removed
   regexp/no-dupe-disjunctions
   37:87   warning  Unexpected useless alternative. This alternative is already covered by '&&?' and can be removed
   regexp/no-dupe-disjunctions
   37:101  warning  Unexpected useless alternative. This alternative is already covered by '\|\|?' and can be removed
   regexp/no-dupe-disjunctions
  187:36   warning  Unexpected useless alternative. This alternative is already covered by '==?' and can be removed
   regexp/no-dupe-disjunctions
  187:81   warning  Unexpected useless alternative. This alternative is already covered by '\d?[<>]&?' and can be removed  regexp/no-dupe-disjunctions
  187:85   warning  Unexpected useless alternative. This alternative is already covered by '\d?[<>]&?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-concurnas.js
  26:51  warning  Unexpected useless alternative. This alternative is already covered by '\?' and can be removed   regexp/no-dupe-disjunctions
p/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-csharp.js
  274:49  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed  regexp/no-dupe-disjunctions
  311:53  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed  regexp/no-dupe-disjunctions
  314:89  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-dhall.js
  62:43  warning  Unexpected useless alternative. This alternative is already covered by '[!=]=' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-django.js
  33:30  warning  Unexpected useless alternative. This alternative is already covered by '[-+*/%=]=?' and can be removed  regexp/no-dupe-disjunctions
  33:38  warning  Unexpected useless alternative. This alternative is already covered by '[-+*/%=]=?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-factor.js
  65:38  warning  Unexpected useless alternative. This alternative is a subset of '\d*\.\d+' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-lilypond.js
  16:72  warning  Unexpected useless alternative. This alternative is already covered by '[^\s()]' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-parigp.js
  28:113  warning  Unexpected useless alternative. This alternative is already covered by '(?: *<)?(?: *=)?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-psl.js
  22:1361  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-qml.js
  6:35  warning  Unexpected useless alternative. This alternative is already covered by '[^\\()[\]{}"'/]' and can be removed  regexp/no-dupe-disjunctions
  6:55  warning  Unexpected useless alternative. This alternative is already covered by '[^\\()[\]{}"'/]' and can be removed  regexp/no-dupe-disjunctions

✖ 25 problems (0 errors, 25 warnings)
```

</details>

The new NFA approach found 28 problems: (This is an apples-to-apples comparison. The NFA approach actually finds more problems _types_ than the AST approach. These additional problem types are not reported here.)

<details>

```
C:\Users\micha\Git\prism\components\prism-abap.js
  19:4043  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-actionscript.js
  2:316  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-agda.js
  22:76  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-bash.js
   37:23   warning  Unexpected useless alternative. This alternative is already covered by '--?' and can be removed
   regexp/no-dupe-disjunctions
   37:32   warning  Unexpected useless alternative. This alternative is already covered by '\+\+?' and can be removed
   regexp/no-dupe-disjunctions
   37:48   warning  Unexpected useless alternative. This alternative is already covered by '\*\*?' and can be removed
   regexp/no-dupe-disjunctions
   37:87   warning  Unexpected useless alternative. This alternative is already covered by '&&?' and can be removed
   regexp/no-dupe-disjunctions
   37:101  warning  Unexpected useless alternative. This alternative is already covered by '\|\|?' and can be removed
   regexp/no-dupe-disjunctions
  187:36   warning  Unexpected useless alternative. This alternative is already covered by '==?' and can be removed
   regexp/no-dupe-disjunctions
  187:81   warning  Unexpected useless alternative. This alternative is already covered by '\d?[<>]&?' and can be removed  regexp/no-dupe-disjunctions
  187:85   warning  Unexpected useless alternative. This alternative is already covered by '\d?[<>]&?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-cmake.js
  17:1198  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions
  17:2326  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions
  17:3870  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions
  17:4791  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-concurnas.js
  26:51  warning  Unexpected useless alternative. This alternative is already covered by '\?' and can be removed   regexp/no-dupe-disjunctions
  28:30  warning  Unexpected useless alternative. This alternative is already covered by '\w*' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-csharp.js
  274:49  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed
 regexp/no-dupe-disjunctions
  311:53  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed
 regexp/no-dupe-disjunctions
  314:89  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed
 regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-dhall.js
  62:43  warning  Unexpected useless alternative. This alternative is already covered by '[!=]=' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-django.js
  33:30  warning  Unexpected useless alternative. This alternative is already covered by '[-+*/%=]=?' and can be removed  regexp/no-dupe-disjunctions
  33:38  warning  Unexpected useless alternative. This alternative is already covered by '[-+*/%=]=?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-lilypond.js
  16:72  warning  Unexpected useless alternative. This alternative is already covered by '[^\s()]' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-parigp.js
  28:113  warning  Unexpected useless alternative. This alternative is already covered by '(?: *<)?(?: *=)?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-psl.js
  22:1361  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-qml.js
  6:35  warning  Unexpected useless alternative. This alternative is already covered by '[^\\()[\]{}"'/]' and can be removed  regexp/no-dupe-disjunctions
  6:55  warning  Unexpected useless alternative. This alternative is already covered by '[^\\()[\]{}"'/]' and can be removed  regexp/no-dupe-disjunctions

✖ 28 problems (0 errors, 28 warnings)
```

</details>

The hybrid approach found 29 problems:

<details>

```
C:\Users\micha\Git\prism\components\prism-abap.js
  19:4043  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-actionscript.js
  2:316  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-agda.js
  22:76  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-bash.js
   37:23   warning  Unexpected useless alternative. This alternative is already covered by '--?' and can be removed
   regexp/no-dupe-disjunctions
   37:32   warning  Unexpected useless alternative. This alternative is already covered by '\+\+?' and can be removed
   regexp/no-dupe-disjunctions
   37:48   warning  Unexpected useless alternative. This alternative is already covered by '\*\*?' and can be removed
   regexp/no-dupe-disjunctions
   37:87   warning  Unexpected useless alternative. This alternative is already covered by '&&?' and can be removed
   regexp/no-dupe-disjunctions
   37:101  warning  Unexpected useless alternative. This alternative is already covered by '\|\|?' and can be removed
   regexp/no-dupe-disjunctions
  187:36   warning  Unexpected useless alternative. This alternative is already covered by '==?' and can be removed
   regexp/no-dupe-disjunctions
  187:81   warning  Unexpected useless alternative. This alternative is already covered by '\d?[<>]&?' and can be removed  regexp/no-dupe-disjunctions
  187:85   warning  Unexpected useless alternative. This alternative is already covered by '\d?[<>]&?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-cmake.js
  17:1198  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions
  17:2326  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions
  17:3870  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions
  17:4791  warning  Unexpected useless alternative. This alternative is a subset of '\w+_(?:CLANG_TIDY|COMPILER_LAUNCHER|CPPCHECK|CPPLINT|INCLUDE_WHAT_YOU_USE|OUTPUT_NAME|POSTFIX|VISIBILITY_PRESET)' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-concurnas.js
  26:51  warning  Unexpected useless alternative. This alternative is already covered by '\?' and can be removed   regexp/no-dupe-disjunctions
  28:30  warning  Unexpected useless alternative. This alternative is already covered by '\w*' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-csharp.js
  274:49  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed
 regexp/no-dupe-disjunctions
  311:53  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed
 regexp/no-dupe-disjunctions
  314:89  warning  Unexpected useless alternative. This alternative is already covered by '[^"'/()]' and can be removed
 regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-dhall.js
  62:43  warning  Unexpected useless alternative. This alternative is already covered by '[!=]=' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-django.js
  33:30  warning  Unexpected useless alternative. This alternative is already covered by '[-+*/%=]=?' and can be removed
  regexp/no-dupe-disjunctions
  33:38  warning  Unexpected useless alternative. This alternative is already covered by '[-+*/%=]=?' and can be removed
  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-factor.js
  65:38  warning  Unexpected useless alternative. This alternative is a subset of '\d*\.\d+' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-lilypond.js
  16:72  warning  Unexpected useless alternative. This alternative is already covered by '[^\s()]' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-parigp.js
  28:113  warning  Unexpected useless alternative. This alternative is already covered by '(?: *<)?(?: *=)?' and can be removed  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-psl.js
  22:1361  warning  Unexpected duplicate alternative  regexp/no-dupe-disjunctions

C:\Users\micha\Git\prism\components\prism-qml.js
  6:35  warning  Unexpected useless alternative. This alternative is already covered by '[^\\()[\]{}"'/]' and can be removed  regexp/no-dupe-disjunctions
  6:55  warning  Unexpected useless alternative. This alternative is already covered by '[^\\()[\]{}"'/]' and can be removed  regexp/no-dupe-disjunctions

✖ 29 problems (0 errors, 29 warnings)
```

</details>

The interesting bit here is the difference between AST and NFA approach. The one additional problem found by the AST approach is unfortunately a bug (#178). The 4 additional problems found by the NFA approach are true positives and were too difficult to detect for the AST approach.

This should demonstrate how powerful the NFA approach is. However, the NFA approach has its downsides too:

1. It can't handle assertions very well. Most of the time, it can't analyze alternatives with assertions at all. This is why the AST approach is still necessary.
2. Performance. 
    | Approach | Time taken |
    | --- | --- |
    | AST | 5s |
    | NFA | 15s |
    | Hybrid | 20s |

RunDevelopment commented on lib/rules/no-dupe-disjunctions.ts:
```diff
@@ -9,6 +10,492 @@ import type {
+            } else {
+                // The else case is interesting.
+                // It means that _some_ of the paths in the current alternative are useless.
+                // TODO: Decide whether this should be reported as well.
```
What do we want to do here?

In this case, the alternative is partially covered by previous alternatives. Should we report this too? E.g. `\d|\w`, `Foo|\w`.

ota-meshi replied:
I think `\d|\w` can delete `\d`, so we may report it. However, I don't think `Foo|\w` can remove one alt. I don't think this rule should be reported. I think it might be nice to have an option to check for partial duplication or another rule.

ota-meshi replied:
Perhaps I misunderstood. Did you say that there is a partial duplication of `(?:\d|\w)|(?:Foo|\w)`?
I think it would be useful to be able to tell the user the correct `\w` location.

RunDevelopment replied:
> I think it might be nice to have an option to check for partial duplication or another rule.

Then let's add an option. Adding another rule for this would also be possible but we would have to do all the NFA operations again.

> Perhaps I misunderstood.

No, you didn't. :)

RunDevelopment replied:
I thought about this some more. `\d|\w` is an interesting case and will be reported.

However, `Foo|\w` will not be reported. I mean, the "fixed" pattern would be `Foo|F(?!oo)|[a-zA-EG-Z0-9_]`. There's no point in reporting this.

RunDevelopment commented on lib/rules/no-dupe-disjunctions.ts:
```diff
@@ -9,6 +10,492 @@ import type {
+    right: ReadonlyNFA,
+): SubsetRelation {
+    try {
+        const inter = DFA.fromIntersection(left, right)
+        inter.minimize()
+
+        const l = DFA.fromFA(left)
+        l.minimize()
+
+        const r = DFA.fromFA(right)
```
DFA constructions can blow up exponentially. Refa detects these exponential blowups and throws an error after a certain point. The default for this point is quite conservative at 10k nodes. For perspective, a DFA construction producing 10k nodes takes about 1 to 10ms, and one producing 100k nodes takes about 10 to 100ms, one producing 1M nodes takes about 100 to 1000ms, and so on.

DFA constructions are necessary to figure out which error message to use. If the DFA construction fails, we currently default to the "overlap" error message (instead of reporting e.g. a subset (=useless) alternative). Maybe we should increase the limit and potentially take more time for more accurate error messages?

ota-meshi replied:
Does this mean there is a possibility of false positives?
I think false negatives are better than false positives, so if we get an error in DFA construction, I think we shouldn't report the error.
I think it is better if you can not limit the number of nodes.

RunDevelopment replied:
No, false positives aren't possible.

> I think it is better if you can not limit the number of nodes.

That's not a good idea. DFA construction blows up exponentially, so even a small regex can create a DFA that is gigabytes in size. I don't want your plugin crashing ESLint by eating up all the available memory :)

I'll increase the node limit to 100k.

ota-meshi replied:
> No, false positives aren't possible.

Does that mean you may not be able to report an easy to understand message?
If so, I don't think there is any problem :+1:

RunDevelopment replied:
It means that we might report the wrong _type_ of message. 

I.e. for `A|B`, `B` might be a subset of `A` but if the DFA construction fails, we will report it as "overlap" instead of "subset".

There won't be any cases, where we tell the user that it's safe to remove something when it isn't. But there might be cases where something is safe to remove but we don't tell the user.

RunDevelopment commented on lib/rules/no-dupe-disjunctions.ts:
```diff
@@ -9,6 +10,492 @@ import type {
+
+                case SubsetRelation.none:
+                case SubsetRelation.unknown:
+                    // TODO: we might want to report cases where the relation
```
As I said in the [other comment](https://github.com/ota-meshi/eslint-plugin-regexp/pull/179/files#r619676707), the DFA construction might fail. If it fails, we don't know which message to display and currentls default to "overlap".

Should we display a special message if this happens?

The problem might be interesting for the user but I don't want to bother them with implementation details.

ota-meshi replied:
I don't think DFA construction errors should report a report.

RunDevelopment replied:
Ok. I'll continue to report them as simple overlaps (if allowed by the user).

RunDevelopment commented:
@ota-meshi Another question.

The NFA approach also finds supersets and general overlap. While users are probably interested in supersets (e.g. `\b(?:Foo|\w+)\b`), I'm not so sure about overlaps.

An overlap occurs if two alternatives can both accept the same string. E.g. the overlap of `a.*|.*b` is `a.*b`.

Overlaps are _usually_ not harmful and hard to remove (e.g. try to remove the overlap in `a.*|.*b`). However, overlap can cause __exponential backtracking__. 

So my idea was to only report overlaps if they cause exponential backtracking. But this doesn't work because people might use patterns to build other patterns. Simple example:

```js
const foo = /(?:ac?|\wb?)a/ // overlap but not exp. backtracking
const bar = RegExp("^(?:" + foo.source + ")+$") // exp backtracking
```

How should we handle this?

ota-meshi commented:
I think we shouldn't consider not using `/(?:ac?|\wb?)a/` as a regular expression. 
Not using regular expression literals as regular expressions is almost always a mistake for the user. Perhaps the user should use a string. This is the same as `String(1234)` and I don't think users should use it.

RunDevelopment commented:
Firstly, let's give these regex literals that aren't used as regexes a name: let's call them partial patterns or partial regexes.

Over at [Prism](https://github.com/PrismJS/prism), we use partial patterns of the form `/regex/.source`, so this problem is somewhat relevant for us. These partial patterns have a few advantages compared to strings (IDE syntax highlighting, no double escaping for backslashes, early syntax errors) which is why I introduced them to Prism. (Our build process inlines the source, so there's no performance overhead for our users.)

So there are valid uses for partial patterns.

ota-meshi commented:
> (IDE syntax highlighting, no double escaping for backslashes, early syntax errors)

I didn't know the practice. That's an interesting practice.
If you think that should be taken into account, do we want to keep track of variables and not report an error if user only use `.source`? (However, if `.source` is done outside the file, it cannot be tracked.)
I also think users can use the `eslint-disable` comment.

RunDevelopment commented:
There are some other rules that also assume regex literals to be used as regexes. E.g. I added the `allowTop` option to `no-useless-non-capturing-group` specifically for partial patterns.

Tracking doesn't have to be perfect. In the case of Prism, simply looking for `.source` would be enough. We can always improve it later.

If it's ok with you, then I'll open an issue for this and implement this rule assuming non-partial regexes.

RunDevelopment commented:
@ota-meshi Bad news. I have found two bugs in refa that have to be fixed before we can merge this rule. Both bugs only affect DFAs, so all other code using refa is fine. I'm currently working on refa v0.8 and will try to release a new version by the end of the week.

RunDevelopment commented:
I think this ready for review now. I implemented everything I wanted.

The question now is: What options should this rule have?

I only added two options for now: `report` and `alwaysReportExponentialBacktracking` (see docs). Do we need more? 

I removed the `disallowNeverMatch` option, so this is a breaking change. Should I re-add it in some form until v1.0?

Also, I'm not sure whether it's a good idea to allow `alwaysReportExponentialBacktracking: false`. We are basically allowing our users to hurt themselves.

ota-meshi commented:
Thank you for lot of work!

> The question now is: What options should this rule have?

I reconsidered it. I think the option should be applied depending on what the user wants to opt out of, but for now I don't think there is any reason to want to opt out, so I don't think we need to offer any more options.
I want we to change to the behavior of `{fastAst: false, noNfa: false,}` that reports the most.

ota-meshi commented on lib/rules/no-dupe-disjunctions.ts:
```diff
@@ -22,106 +630,178 @@ export default createRule("no-dupe-disjunctions", {
             {
                 type: "object",
                 properties: {
-                    disallowNeverMatch: { type: "boolean" },
```
I want to leave `disallowNeverMatch` in the schema and remove it in v1.0. We don't have to handle `disallowNeverMatch`.

RunDevelopment replied:
Good idea. Should I log a warning if the user tries to use it?

ota-meshi replied:
I don't think we need log. Users who used or did not use `disallowNeverMatch` will be able to use the enhanced `no-dupe-disjunctions` in the next minor version.
By removing the option from the schema in the v1.0 release, users who use the option in v1.0 will be warned.

ota-meshi commented on lib/rules/no-dupe-disjunctions.ts:
```diff
@@ -22,106 +630,180 @@ export default createRule("no-dupe-disjunctions", {
+                    parentNode.alternatives,
+                    regexpContext,
+                    {
+                        fastAst: true,
```
Should we change it to `false` to report many warnings?

```suggestion
                        fastAst: false,
```

RunDevelopment replied:
Yes. My bad.

RunDevelopment replied:
I re-enabled the AST approach and added a test case the NFA approach definitely can't detect.

ota-meshi reviewed (approved):
Thank you for your wonderful work!
I will release a minor version today or tomorrow that includes this change.
